KAFKA-7072: clean up segments only after they expire

vvcephei opened the pull request:
Significant refactor of Segments to use stream-time as the basis of segment expiration.
Previously Segments assumed that the current record time was representative of stream time.

In the event of a "future" event (one whose record time is greater than the stream time), this
would inappropriately drop live segments. Now, Segments will provision the new segment
to house the future event and drop old segments only after they expire.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

vvcephei committed:
KAFKA-7072: clean up segments only after they expire

vvcephei commented:
@guozhangwang @mjsax @bbejeck ,

I'd appreciate a review on this when you get the chance.

Thanks,
-John

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java:
```diff
@@ -97,7 +98,7 @@ public void remove(final Bytes key) {
     public void put(final Bytes key, final byte[] value) {
         final long segmentId = segments.segmentId(keySchema.segmentTimestamp(key));
-        final Segment segment = segments.getOrCreateSegment(segmentId, context);
+        final Segment segment = segments.getOrCreateSegmentIfLive(segmentId, context);
```
nit: maybe add a debug level log4j indicating the put operation is skipped?

vvcephei replied:
I can do that. FWIW, we always skipped such records, I just changed the method name to reflect that fact.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java:
```diff
@@ -121,89 +130,66 @@ void openExisting(final ProcessorContext context) {
-                } catch (InvalidStateStoreException ise) {
-                    // segment may have been closed by streams thread;
-                }
+        final List<Segment> result = new ArrayList<>();
```
This is not necessarily part of this PR, but right now we only purge segments in `put`, should we consider doing purge in `segments` and `allSegments()` as well? They cover all range queries, i.e. both read and writes can trigger segment purging.

vvcephei replied:
We could do that, but it might be better to only do it in `put`.

The disk usage can only grow during `put`, so there's no danger of running out of disk if you garbage collect only during `put`. Collecting in extra places would just be an optimization to keep the disk footprint as small as possible.

Also, it might be surprising to perform disk i/o during a metadata call to list segments.

That said, it wouldn't be incorrect to do it...

guozhangwang replied:
That is a good point. Let's keep our current behavior as only purge on puts then so that range queries do not see surprise perf variance due to rocksdb closing (each may take a few seconds in worst case).

guozhangwang reviewed (commented):
The change lgtm.

A meta comment: did we add any new unit test to expose the original issue? i.e. the test would fail before this PR, and will succeed with this PR.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java:
```diff
@@ -119,11 +120,11 @@ public String name() {
         keySchema.init(ProcessorStateManager.storeChangelogTopic(context.applicationId(), root.name()));
 
-        segments.openExisting(context);
+        segments.openExisting((InternalProcessorContext) context);
```
nit: pass `this.context` to avoid casting

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java:
```diff
@@ -39,7 +40,7 @@
         this.name = name;
         this.keySchema = keySchema;
-        this.segments = new Segments(name, retention, numSegments);
+        this.segments = new Segments(name, retention, Segments.segmentInterval(retention, numSegments));
```
Why this change? The existing code computes segmentInterval internally already. Seems to be better contained, as we pass in `rentention` time anyway?

vvcephei replied:
Actually, this computation was already exposed as a package-private static method, so I'd argue there's no less containment. This way, it's clear that Segments doesn't care about the number of segments, only the retention period and segment interval.

Now that I'm looking at it, that `segmentInterval` doesn't need to live in `Segments`. It's only used in the neighborhood of `RocksDBSementedBytesStore`, so I'll move it up.

vvcephei replied:
Oof, while doing this I discovered a bug in which the SegmentedCache takes a segmentInterval argument, but the WindowBytesStoreSupplier provides instead the *number of segments* to it.

Here's the call passing #segments: https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/state/internals/WindowStoreBuilder.java#L55

Here's the receiving end of the call: https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/state/internals/CachingWindowStore.java#L59

The types are different, but it's int -> long, so Java automatically up-casts it. :(

There's no way to fix that bug without altering the public interface. I guess the right thing to do would be to deprecate https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/state/WindowBytesStoreSupplier.java#L37 and add a `segmentInterval` method instead. 

It's ironic because there is exactly one call to that method in the entire codebase, and it uses it wrong :/

What do you think I should do about this?

vvcephei replied:
I considered deviously just "correcting" the JavaDoc to state that `segments` is the segment interval, but I'm afraid that external code might either be implementing this interface or using it (for a custom store impl)...

vvcephei replied:
I've added a fix for this to this issue (https://github.com/apache/kafka/pull/5253/commits/1d0359614b5444230cd99e7c0e02a0978da81435), but I should probably file a jira just for this issue and write a small KIP proposing the interface change and then create a separate PR just containing https://github.com/apache/kafka/pull/5253/commits/1d0359614b5444230cd99e7c0e02a0978da81435.

WDYT? @mjsax @guozhangwang @bbejeck

guozhangwang replied:
Squinting at the `WindowBytesStoreSupplier` now.. with the current API we are actually enforcing any customized Window bytes store in DSL's `Materialized` to follow the same segmented pattern as `RocksDbWindowBytesStoreSupplier`, right? Ideally a customized window store supplier only need to expose the window size, and maybe retention period. Of course it is a general issue that may need a separate discussion...

For now I think adding retention interval and deprecating num.segments in `WindowBytesStoreSupplier` is okay, but it would require a KIP.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java:
```diff
@@ -67,7 +67,7 @@
     public void setUp() {
         keySchema = new WindowKeySchema();
-        final int retention = 30000;
+        final int retention = 60000;
```
Why this?

vvcephei replied:
The tests incorrectly depended on windows being present after their retention time passed.

mjsax commented on streams/src/test/java/org/apache/kafka/test/InternalMockProcessorContext.java:
```diff
@@ -57,6 +58,7 @@
     private final Map<String, StateStore> storeMap = new LinkedHashMap<>();
     private final Map<String, StateRestoreCallback> restoreFuncs = new HashMap<>();
     private final ToInternal toInternal = new ToInternal();
+    private final AtomicLong streamTimeSupplier = new AtomicLong();
```
Why not `long`?

vvcephei replied:
¯\_(ツ)_/¯

I initially used a "supplier" interface, which AtomicLong matches, but I wound up with a different implementation. I'll change it to `long`.

mjsax reviewed (commented):
Some minor comments.

vvcephei commented:
@guozhangwang I did write a test for it, but wound up losing it during refactoring/merging. You're right, we need one.

vvcephei committed:
BUGFIX: Use segmentInterval internally instead of numSegments

I discovered a bug in which the SegmentedCache takes a segmentInterval
argument, but the WindowBytesStoreSupplier provides instead the number
of segments to it.

Here's the call passing #segments:
https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/state/internals/WindowStoreBuilder.java#L55

Here's the receiving end of the call:
https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/state/internals/CachingWindowStore.java#L59

The types are different, but it's int -> long, so Java automatically
up-casts it. :(

There's no way to fix that bug without altering the public interface. I
guess the right thing to do would be to deprecate
https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/state/WindowBytesStoreSupplier.java#L37
and add a segmentInterval method instead.

vvcephei committed:
debug log

vvcephei committed:
add missing unit test for future events

vvcephei committed:
cr comment

vvcephei committed:
switch from AtomicLong to primitive long

vvcephei commented:
@guozhangwang @mjsax @bbejeck ,

I've updated the PR to address all the comments (and responded to the one about garbage collecting outside of `poll()`).

I discovered one issue while looking into a comment from Matthias on "hoisting" the `segmentInterval` out of `Segments`. It turns out that CachingWindowStore actually also needs `segmentInterval`, but we were passing in number-of-segments. This PR fixes it, but it's a public interface change.

The conversation is hidden now, so I've copied it below.

> mjsax 12 hours ago  Member
> Why this change? The existing code computes segmentInterval internally already. Seems to be better contained, as we pass in rentention time anyway?
> 
> 
> vvcephei an hour ago  Contributor
> Actually, this computation was already exposed as a package-private static method, so I'd argue there's no less containment. This way, it's clear that Segments doesn't care about the number of segments, only the retention period and segment interval.
> 
> Now that I'm looking at it, that segmentInterval doesn't need to live in Segments. It's only used in the neighborhood of RocksDBSementedBytesStore, so I'll move it up.
> 
> 
> vvcephei an hour ago  Contributor
> Oof, while doing this I discovered a bug in which the SegmentedCache takes a segmentInterval argument, but the WindowBytesStoreSupplier provides instead the number of segments to it.
> 
> Here's the call passing #segments: https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/state/internals/WindowStoreBuilder.java#L55
> 
> Here's the receiving end of the call: https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/state/internals/CachingWindowStore.java#L59
> 
> The types are different, but it's int -> long, so Java automatically up-casts it. :(
> 
> There's no way to fix that bug without altering the public interface. I guess the right thing to do would be to deprecate https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/state/WindowBytesStoreSupplier.java#L37 and add a segmentInterval method instead.
> 
> It's ironic because there is exactly one call to that method in the entire codebase, and it uses it wrong :/
> 
> What do you think I should do about this?
> 
> 
> vvcephei an hour ago  Contributor
> I considered deviously just "correcting" the JavaDoc to state that segments is the segment interval, but I'm afraid that external code might either be implementing this interface or using it (for a custom store impl)...
> 
> 
> vvcephei 12 minutes ago  • 
>  Contributor
> I've added a fix for this to this issue (1d03596), but I should probably file a jira just for this issue and write a small KIP proposing the interface change and then create a separate PR just containing 1d03596.
> 
> WDYT? @mjsax @guozhangwang @bbejeck

vvcephei committed:
unused import

vvcephei committed:
revert bugfix and interface change, pending KIP

vvcephei commented:
Ok, I've reverted the interface change from the current PR, preserving the current (incorrect) behavior.

I'm not sure I want to piggy-back it on final-update, though, because it's unrelated. I'll just make a small, hopefully uncontroversial KIP to fix this interface.

vvcephei committed:
Update RocksDbWindowBytesStoreSupplier.java

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/Stores.java:
```diff
@@ -161,13 +161,15 @@ public static WindowBytesStoreSupplier persistentWindowStore(final String name,
-        if (numSegments < 1) {
-            throw new IllegalArgumentException("numSegments cannot must smaller than 1");
+        if (numSegments < 2) {
+            throw new IllegalArgumentException("numSegments cannot must smaller than 2");
```
Previously, the `RocksDbWindowBytesStoreSupplier` constructor enforced the minimum of 2, so if you passed `< 1`, you'd get an exception from here, and if you passed `<2`, you'd get the exception from the constructor called just below here.

This way, we collapse both checks into one equivalent check here.

vvcephei commented:
@tedyu @guozhangwang , I've created KIP-319 to propose that change.

Otherwise, I think this PR is ready for final reviews!

guozhangwang commented:
LGTM. waiting for @bbejeck or @mjsax to have another pass before merging.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDbSessionBytesStoreSupplier.java:
```diff
@@ -55,8 +53,6 @@ public String metricsScope() {
-        return Segments.segmentInterval(
-            retentionPeriod,
-            NUM_SEGMENTS);
+        return Math.max(retentionPeriod / 2, 60_1000L);
```
`60_1000L` -> `60_000L` ? (originally it was `MIN_SEGMENT_INTERVAL = 60 * 1000L`)

vvcephei replied:
oh! Thank you!

mjsax reviewed (commented):
One comment. Otherwise LGTM.

vvcephei committed:
Update RocksDbSessionBytesStoreSupplier.java

guozhangwang commented:
Merged to trunk.
